SwiftCompilerSources: update `Package.swift`

compnerd opened the pull request:
Update the package manager manifest to properly enable C++ interoperability.  This allows us to start trying to build SwiftCompilerSources with SPM again.

compnerd committed:
SwiftCompilerSources: set the C++ language to C++17

LLVM, clang, and Swift use C++17.  Explicitly specify the C++ language
standard to use.

compnerd committed:
SwiftCompilerSources: mark the required C++ interoperability

Mark the Swift flags for interoperability mode rather than using the old
spelling for enabling C++ interoperability.  This is important as it
allows the `cxxLanguageStandard` to flow through to the clang importer
which is required to enable the use of the C++ headers.

eeckstein reviewed (approved):
lgtm

compnerd commented:
@swift-ci please test

hyp commented on SwiftCompilerSources/Package.swift:
```diff
@@ -15,6 +15,7 @@ import PackageDescription
   static let defaultSwiftSettings: [SwiftSetting] = [
+    .interoperabilityMode(.Cxx),
     .unsafeFlags([
       "-Xfrontend", "-validate-tbd-against-ir=none",
```
you can remove these flags, they're set now for interop.

compnerd replied:
Thanks @hyp!

compnerd committed:
SwiftCompilerSources: remove extraneous flags

Remove the flags that are now implicitly setup by SPM.  Addresses
feedback from @hyp.

compnerd committed:
SwiftCompilerSources: modernise and inline default flags

Use the `cxxSettings` to pass along header search paths.  With those
removed, we are simply adding the interoperability mode and unsafe
flags, which we can inline.  This simplifies the package definition a
bit.

compnerd committed:
Merge branch 'main' into interoperation

compnerd commented:
@swift-ci please smoke test

compnerd commented:
@swift-ci please smoke test macOS platform
